client.c

```diff
@@ -1140,37 +1140,30 @@ static void handle_gs(struct fio_client *client, struct fio_net_cmd *cmd)
 static void handle_job_opt(struct fio_client *client, struct fio_net_cmd *cmd)
 {
 	struct cmd_job_option *pdu = (struct cmd_job_option *) cmd->payload;
-	struct print_option *p;
-
-	if (!job_opt_object)
-		return;
 
 	pdu->global = le16_to_cpu(pdu->global);
 	pdu->truncated = le16_to_cpu(pdu->truncated);
 	pdu->groupid = le32_to_cpu(pdu->groupid);
 
-	p = malloc(sizeof(*p));
-	p->name = strdup((char *) pdu->name);
-	if (pdu->value[0] != '\0')
-		p->value = strdup((char *) pdu->value);
-	else
-		p->value = NULL;
-
 	if (pdu->global) {
-		const char *pos = "";
+		if (!job_opt_object)
+			return;
 
-		if (p->value)
-			pos = p->value;
-
-		json_object_add_value_string(job_opt_object, p->name, pos);
+		json_object_add_value_string(job_opt_object,
+					     strdup((char *)pdu->name) ? : "",
```

Comment:
If we want to match was was there before don't we want to pass NULL if duplication of `pdu->name` couldn't allocate memory?

Reply:
I'm left wondering how the duplicated name could ever be freed because it might have been dynamically allocated or it might be constant (if it is the empty string we pass here). Having said that, this is how the code was before so I guess there's no harm?

Reply:
> If we want to match was was there before don't we want to pass NULL if duplication of `pdu->name` couldn't allocate memory?

I think not passing NULL to the `name` argument of `json_object_add_value_string()` is a bug fix. The `name` argument is passed indirectly to `strdup()` and the `strdup()` argument must not be NULL.

Reply:
> I'm left wondering how the duplicated name could ever be freed because it might have been dynamically allocated or it might be constant (if it is the empty string we pass here). Having said that, this is how the code was before so I guess there's no harm?

Is this perhaps the code that frees the memory allocated by the code touched by this patch?
```
	json_free_object(root);
```

Reply:
@bvanassche My guess is that it it would be done by `json_free_value()` over in https://github.com/axboe/fio/blob/master/json.c#L155 as that is (eventually) called by functions called from `json_free_object()`. But again how is that meant to handle the constant string `""`? Do we simply never try and free these objects?

Reply:
> I think not passing NULL to the `name` argument of `json_object_add_value_string()` is a bug fix. The `name` argument is passed indirectly to `strdup()` and the `strdup()` argument must not be NULL.

Fair enough. Tracing through `strdup_escape()` shows it has a call to `strlen()` early on and that call expects its argument to not be NULL.

Reply:
> @bvanassche My guess is that it it would be done by `json_free_value()` over in https://github.com/axboe/fio/blob/master/json.c#L155 as that is (eventually) called by functions called from `json_free_object()`. But again how is that meant to handle the constant string `""`? Do we simply never try and free these objects?

Please take a look at the new patch series that has been pushed. An additional memory leak has been fixed.